File: tests/cross_driver/node/sessions_smoke.js
// Cross-driver logical-sessions smoke — Node (mongo-node-driver).
//
// Exercises startSession / endSessions / refreshSessions through the
// raw runCommand surface (the driver's high-level startSession() API
// keeps the lsid private, so we drive the wire commands directly to
// verify the registry state our handlers maintain).

const { MongoClient, Binary } = require('mongodb');

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

async function main() {
  const uri = process.env.MONGODB_URI;
  if (!uri) fail('MONGODB_URI not set');

  const client = new MongoClient(uri, { serverSelectionTimeoutMS: 5000 });
  await client.connect();
  try {
    const adminDb = client.db('admin');

    // 1. startSession returns {id: BinData(4, uuid), timeoutMinutes}.
    const started = await adminDb.command({ startSession: 1 });
    if (!started || started.ok !== 1) fail(`startSession failed: ${JSON.stringify(started)}`);
    const wrapped = started.id;
    if (!wrapped || !(wrapped.id instanceof Binary)) {
      fail(`startSession id shape: ${JSON.stringify(started)}`);
    }
    if (wrapped.id.sub_type !== 4) fail(`lsid sub_type: ${wrapped.id.sub_type}`);
    const lsidBytes = wrapped.id.value();
    if (lsidBytes.length !== 16) fail(`lsid length: ${lsidBytes.length}`);
    if (started.timeoutMinutes !== 30) {
      fail(`timeoutMinutes: ${started.timeoutMinutes}, want 30`);
    }

    // 2. endSessions on the freshly-minted lsid → {ok: 1}.
    const ended = await adminDb.command({ endSessions: [wrapped] });
    if (ended.ok !== 1) fail(`endSessions: ${JSON.stringify(ended)}`);

    // 3. refreshSessions implicit-creates an unknown lsid; the server
    // accepts arbitrary 16-byte UUIDs and returns ok.
    const fakeLsid = Buffer.from('0123456789abcdef', 'utf8');
    const fakeWrapped = { id: new Binary(fakeLsid, 4) };
    const refreshed = await adminDb.command({ refreshSessions: [fakeWrapped] });
    if (refreshed.ok !== 1) fail(`refreshSessions: ${JSON.stringify(refreshed)}`);

    console.log('OK');
  } finally {
    await client.close();
  }
}

main().catch((err) => fail(`uncaught: ${err && err.stack ? err.stack : err}`));

File: tests/cross_driver/java/src/main/java/com/secantus/smokes/SessionsSmoke.java
package com.secantus.smokes;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoDatabase;
import org.bson.Document;
import org.bson.types.Binary;

import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Cross-driver logical-sessions smoke — Java.
 * <p>
 * Drives ``startSession`` / ``endSessions`` / ``refreshSessions`` via
 * raw {@code runCommand} so the wire-level lsid round-trip is what's
 * exercised (not the driver's high-level {@code ClientSession} API).
 */
public final class SessionsSmoke {

    private SessionsSmoke() {}

    public static void main(String[] args) {
        String uri = System.getenv("MONGODB_URI");
        if (uri == null) {
            System.err.println("MONGODB_URI not set");
            System.exit(2);
        }

        MongoClientSettings settings = MongoClientSettings.builder()
            .applyConnectionString(new ConnectionString(uri))
            .applyToClusterSettings(b -> b.serverSelectionTimeout(30, TimeUnit.SECONDS))
            .build();

        try (MongoClient client = MongoClients.create(settings)) {
            MongoDatabase admin = client.getDatabase("admin");

            // 1. startSession returns {id: BinData(4, uuid), timeoutMinutes}.
            Document started = admin.runCommand(new Document("startSession", 1));
            if (started.getDouble("ok") != 1.0) {
                System.err.printf("startSession failed: %s%n", started.toJson());
                System.exit(1);
            }
            Document idWrap = started.get("id", Document.class);
            if (idWrap == null) {
                System.err.printf("startSession id wrapper missing: %s%n", started.toJson());
                System.exit(1);
            }
            Binary lsidBin = (Binary) idWrap.get("id");
            if (lsidBin == null || lsidBin.getType() != 4 || lsidBin.getData().length != 16) {
                System.err.printf("lsid shape: %s%n", lsidBin);
                System.exit(1);
            }
            Integer timeout = started.getInteger("timeoutMinutes");
            if (timeout == null || timeout != 30) {
                System.err.printf("timeoutMinutes: %s, want 30%n", timeout);
                System.exit(1);
            }

            // 2. endSessions on the freshly-minted lsid.
            Document ended = admin.runCommand(new Document("endSessions",
                List.of(new Document("id", lsidBin))));
            if (ended.getDouble("ok") != 1.0) {
                System.err.printf("endSessions: %s%n", ended.toJson());
                System.exit(1);
            }

            // 3. refreshSessions implicit-creates an unknown lsid.
            Binary fakeLsid = new Binary((byte) 4, "0123456789abcdef".getBytes());
            Document refreshed = admin.runCommand(new Document("refreshSessions",
                List.of(new Document("id", fakeLsid))));
            if (refreshed.getDouble("ok") != 1.0) {
                System.err.printf("refreshSessions: %s%n", refreshed.toJson());
                System.exit(1);
            }

            System.out.println("OK");
        }
    }
}
